flang/lib/Lower/ConvertType.cpp

```diff
@@ -410,20 +409,12 @@ class TypeBuilder {
           auto &p{*param};
           ps.push_back(std::pair{p.name().ToString(), gen(p)});
         }
-#if 0
-        // this functionality is missing in the front-end
-        for (auto &comp : details.componentDecls()) {
-          auto &c{*comp};
-          cs.push_back(std::pair{c.name().ToString(), gen(c)});
-        }
-#else
-        emitWarning("the front-end returns symbols of derived type that have "
-                    "components that are simple names and not symbols, so "
-                    "cannot construct type " +
-                    toStringRef(symbol.name()));
-#endif
+        emitError("the front-end returns symbols of derived type that have "
+                  "components that are simple names and not symbols, so cannot "
+                  "construct the type '" +
+                  toStringRef(symbol.name()) + "'");
```

Comment:
Going through components is tricky, there is a helper for that: `Fortran::semantics::ComponentIterator`.
https://github.com/flang-compiler/f18-llvm-project/blob/d4579b7ef696dc932f3174896ee5de7d235c6462/flang/include/flang/Semantics/tools.h#L350
It should be possible to write something like:
```
auto components = Fortran::semantics::ComponentIterator<Fortran::semantics::ComponentKind::Ultimate>(type->derivedTypeSpec());
for (const auto& component : components) {
 // If I remember well, component should be const Symbol& of an ultimate component.
}
```
Note: I am not sure which component the iterator should go through here though ? Ultimate or Direct or something else ?

Anyway, this is off-topic for this PR, but maybe the message should be transformed into a TODO to use this.

Reply:
It may well have improved since I first wrote this code and error message. The symbol table used to just keep raw `CharBlock` values IIRC, which were not useful when one needed the actual symbol.